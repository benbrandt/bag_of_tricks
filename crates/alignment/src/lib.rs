#![deny(clippy::all)]
#![warn(clippy::pedantic)]

use std::{convert::TryFrom, f64::consts::E, fmt};

use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

fn exp_weight(val: usize) -> f64 {
    E.powi(i32::try_from(val).unwrap_or_default())
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

impl Attitude {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Morality {
    Evil,
    Good,
    Neutral,
}

impl Morality {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

/// Trait to describe how this entity affects a character's alignment
pub trait AlignmentInfluences {
    /// List of attitude influences
    fn attitude(&self) -> Vec<Attitude> {
        vec![]
    }

    /// List of morality influences
    fn morality(&self) -> Vec<Morality> {
        vec![]
    }
}

/// Character alignment, both attitude and morality
#[derive(Clone, Deserialize, Serialize)]
pub struct Alignment(pub Attitude, pub Morality);

impl Alignment {
    /// Generate alignment, weighted by influences from other choices on the character sheet
    ///
    /// # Panics
    ///
    /// Will panic if weighting logic is wrong or there are no attitudes/moralities at all
    pub fn gen(
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> Self {
        let attitude = *Attitude::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(attitude_influences))
            .unwrap();
        let morality = *Morality::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(morality_influences))
            .unwrap();
        Self(attitude, morality)
    }

    /// Weight of a particular alignment based on influences.
    /// Useful for comparing things like deities.
    #[must_use]
    pub fn weight(
        &self,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> f64 {
        self.0.weight(attitude_influences) + self.1.weight(morality_influences)
    }
}

impl AlignmentInfluences for Alignment {
    fn attitude(&self) -> Vec<Attitude> {
        vec![self.0]
    }

    fn morality(&self) -> Vec<Morality> {
        vec![self.1]
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
