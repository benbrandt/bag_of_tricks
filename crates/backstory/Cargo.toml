[package]
name = "backstory"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
