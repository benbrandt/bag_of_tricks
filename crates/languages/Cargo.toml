[package]
name = "languages"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
deities = { path = "../deities" }
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
