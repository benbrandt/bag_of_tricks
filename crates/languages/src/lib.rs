use deities::{Pantheon, PantheonWeight, Pantheons};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum LanguageType {
    Exotic,
    Standard,
}

impl LanguageType {
    fn weight(self) -> f64 {
        match self {
            Self::Exotic => 1.0,
            Self::Standard => 2.0,
        }
    }
}

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gith,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub fn gen(
        rng: &mut impl Rng,
        current_languages: &[Language],
        (amount, language_type): (usize, Option<LanguageType>),
    ) -> Vec<Self> {
        language_type
            .map_or_else(
                || Self::iter().collect::<Vec<Self>>(),
                |lt| {
                    let filtered = Self::iter()
                        .filter(|l| l.language_type() == lt)
                        .collect::<Vec<Self>>();
                    if filtered.is_empty() {
                        Self::iter().collect::<Vec<Self>>()
                    } else {
                        filtered
                    }
                },
            )
            .iter()
            .filter(|l| !current_languages.contains(l))
            .copied()
            .collect::<Vec<_>>()
            .choose_multiple_weighted(rng, amount, |l| l.language_type().weight())
            .unwrap()
            .copied()
            .collect()
    }

    fn language_type(self) -> LanguageType {
        match self {
            Self::Common
            | Self::Dwarvish
            | Self::Elvish
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Orc => LanguageType::Standard,
            Self::Abyssal
            | Self::Celestial
            | Self::DeepSpeech
            | Self::Draconic
            | Self::Gith
            | Self::Infernal
            | Self::Primordial
            | Self::Sylvan
            | Self::Undercommon => LanguageType::Exotic,
        }
    }
}

impl Pantheons for Language {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        (match self {
            Language::Abyssal
            | Language::Celestial
            | Language::DeepSpeech
            | Language::Gith
            | Language::Infernal
            | Language::Primordial
            | Language::Sylvan => vec![],
            Language::Common => vec![Pantheon::ForgottenRealms],
            Language::Draconic => vec![Pantheon::Dragon, Pantheon::Kobold, Pantheon::Lizardfolk],
            Language::Dwarvish => vec![Pantheon::Dwarven],
            Language::Elvish => vec![Pantheon::Elven],
            Language::Giant => vec![Pantheon::Giant],
            Language::Gnomish => vec![Pantheon::Gnomish],
            Language::Goblin => vec![Pantheon::Bugbear, Pantheon::Goblin],
            Language::Halfling => vec![Pantheon::Halfling],
            Language::Orc => vec![Pantheon::Orc],
            Language::Undercommon => vec![Pantheon::Drow, Pantheon::Duergar],
        })
        .into_iter()
        .map(|p| (p, PantheonWeight::Possible))
        .collect::<Vec<_>>()
    }
}

/// Trait for describing what languages are provided by a given object.
pub trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (0, None)
    }
}
