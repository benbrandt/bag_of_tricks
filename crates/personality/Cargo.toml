[package]
name = "personality"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
alignment = { path = "../alignment" }
serde = { version = "1", features = ["derive"] }
strum = { version = "0.25", features = ["derive"] }
