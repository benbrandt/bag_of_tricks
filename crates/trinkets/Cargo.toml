[package]
name = "trinkets"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1", features = ["derive"] }
