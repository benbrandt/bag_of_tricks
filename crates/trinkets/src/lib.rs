use serde::{Deserialize, Serialize};

#[derive(Clone, Deserialize, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub enum TrinketOption {
    Default,
    Elven,
    Goblin,
    Gothic,
    Custom(Vec<String>),
}

impl TrinketOption {
    pub fn trinkets(&self) -> Vec<String> {
        match self {
            Self::Default => TRINKETS.iter().map(|&t| t.to_string()).collect(),
            Self::Elven => ELVEN_TRINKETS.iter().map(|&t| t.to_string()).collect(),
            Self::Goblin => GOBLIN_STATUS_SYMBOLS
                .iter()
                .map(|&t| t.to_string())
                .collect(),
            Self::Gothic => GOTHIC_TRINKETS.iter().map(|&t| t.to_string()).collect(),
            Self::Custom(l) => l.clone(),
        }
    }
}

pub trait Trinkets {
    fn trinket_options(&self) -> Vec<TrinketOption> {
        vec![]
    }
}

const TRINKETS: &[&str] = &[
    "A mummified goblin hand",
    "A piece of crystal that faintly glows in the moonlight",
    "A gold coin minted in an unknown land",
    "A diary written in a language you don't know",
    "A brass ring that never tarnishes",
    "An old chess piece made from glass",
    "A pair of knucklebone dice, each with a skull symbol on the side that would normally show six pips",
    "A small idol depicting a nightmarish creature that gives you unsettling dreams when you sleep near it",
    "A rope necklace from which dangles four mummified elf fingers",
    "The deed for a parcel of land in a realm unknown to you",
    "A 1-ounce block made from an unknown material",
    "A small cloth doll skewered with needles",
    "A tooth from an unknown beast",
    "An enormous scale, perhaps from a dragon",
    "A bright green feather",
    "An old divination card bearing your likeness",
    "A glass orb filled with moving smoke",
    "A 1-pound egg with a bright red shell",
    "A pipe that blows bubbles",
    "A glass jar containing a weird bit of flesh floating in pickling fluid",
    "A tiny gnome-crafted music box that plays a song you dimly remember from your childhood",
    "A small wooden statuette of a smug halfling",
    "A brass orb etched with strange runes",
    "A multicolored stone disk",
    "A tiny silver icon of a raven",
    "A bag containing forty-seven humanoid teeth, one of which is rotten",
    "A shard of obsidian that always feels warm to the touch",
    "A dragon's bony talon hanging from a plain leather necklace",
    "A pair of old socks",
    "A blank book whose pages refuse to hold ink, chalk, graphite, or any other substance or marking",
    "A silver badge in the shape of a five-pointed star",
    "A knife that belonged to a relative",
    "A glass vial filled with nail clippings",
    "A rectangular metal device with two tiny metal cups on one end that throws sparks when wet",
    "A white, sequined glove sized for a human",
    "A vest with one hundred tiny pockets",
    "A small, weightless stone block",
    "A tiny sketch portrait of a goblin",
    "An empty glass vial that smells of perfume when opened",
    "A gemstone that looks like a lump of coal when examined by anyone but you",
    "A scrap of cloth from an old banner",
    "A rank insignia from a lost legionnaire",
    "A tiny silver bell without a clapper",
    "A mechanical canary inside a gnome-crafted lamp",
    "A tiny chest carved to look like it has numerous feet on the bottom",
    "A dead sprite inside a clear glass bottle",
    "A metal can that has no opening but sounds as if it is filled with liquid, sand, spiders, or broken glass (your choice)",
    "A glass orb filled with water, in which swims a clockwork goldfish",
    "A silver spoon with an M engraved on the handle",
    "A whistle made from gold-colored wood",
    "A dead scarab beetle the size of your hand",
    "Two toy soldiers, one with a missing head",
    "A small box filled with different-sized buttons",
    "A candle that can't be lit",
    "A tiny cage with no door",
    "An old key",
    "An indecipherable treasure map",
    "A hilt from a broken sword",
    "A rabbit's foot",
    "A glass eye",
    "A cameo carved in the likeness of a hideous person",
    "A silver skull the size of a coin",
    "An alabaster mask",
    "A pyramid of sticky black incense that smells very bad",
    "A nightcap that, when worn, gives you pleasant dreams",
    "A single caltrop made from bone",
    "A gold monocle frame without the lens",
    "A 1-inch cube, each side painted a different color",
    "A crystal knob from a door",
    "A small packet filled with pink dust",
    "A fragment of a beautiful song, written as musical notes on two pieces of parchment",
    "A silver teardrop earring made from a real teardrop",
    "The shell of an egg painted with scenes of human misery in disturbing detail",
    "A fan that, when unfolded, shows a sleeping cat",
    "A set of bone pipes",
    "A four-leaf clover pressed inside a book discussing manners and etiquette",
    "A sheet of parchment upon which is drawn a complex mechanical contraption",
    "An ornate scabbard that fits no blade you have found so far",
    "An invitation to a party where a murder happened",
    "A bronze pentacle with an etching of a rat's head in its center",
    "A purple handkerchief embroidered with the name of a powerful archmage",
    "Half of a floorplan for a temple, castle, or some other structure",
    "A bit of folded cloth that, when unfolded, turns into a stylish cap",
    "A receipt of deposit at a bank in a far-flung city",
    "A diary with seven missing pages",
    "An empty silver snuffbox bearing an inscription on the surface that says \"dreams\"",
    "An iron holy symbol devoted to an unknown god",
    "A book that tells the story of a legendary hero's rise and fall, with the last chapter missing",
    "A vial of dragon blood",
    "An ancient arrow of elven design",
    "A needle that never bends",
    "An ornate brooch of dwarven design",
    "An empty wine bottle bearing a pretty label that says, \"The Wizard of Wines Winery, Red Dragon Crush, 331422-W\"",
    "A mosaic tile with a multicolored, glazed surface",
    "A petrified mouse",
    "A black pirate flag adorned with a dragon's skull and crossbones",
    "A tiny mechanical crab or spider that moves about when it;s not being observed",
    "A glass jar containing lard with a label that reads, \"Griffon Grease\"",
    "A wooden box with a ceramic bottom that holds a living worm with a head on each end of its body",
    "A metal urn containing the ashes of a hero",
];

const ELVEN_TRINKETS: &[&str] = &[
    "A small notebook that causes anything written in it to disappear after 1 hour",
    "A crystal lens made of ivory and gold that causes anything observed through it to appear to be surrounded by motes of multicolored light",
    "A small golden pyramid inscribed with elven symbols and about the size of a walnut",
    "A cloak pin made from enamel in the shape of a butterfly; when you take the pin off, it turns into a real butterfly, and returns when you are ready to put your cloak back on again",
    "A golden compass that points toward the nearest portal to the Feywild within 10 miles",
    "A small silver spinning top that, when spun, endlessly spins until interrupted",
    "A small songbird made of enamel, gold wire, and precious stone; uttering the songbird's name in Elvish causes the trinket to emit that bird's birdsong",
    "A small enamel flower that, when put in one's hair, animates, tying back the wearer's hair with a living vine with flowers; plucking a single flower from this vine returns it to its inanimate form",
];

const GOBLIN_STATUS_SYMBOLS: &[&str] = &[
    "Earrings and notches in an ear",
    "Rib bones tied into hair",
    "A belt made from raccoon pelts",
    "A gnome's boot used as a hat",
    "A pouch of toenail clippings from an allied ogre",
    "A frog kept in a jar",
    "Fragile helmets made from axe beak eggs",
    "Nose rings",
    "Painted or stained hands",
    "Bugs kept in a bag for snacking",
    "War cry tatooed on chest",
    "Shields made from ankheg chitin",
    "Bracelet made of pieces of goblins turned into stone",
    "Special breed of rat kept as pet",
    "Teeth pulled out in certain places",
    "Owlbear-feather cloaks",
    "Scars from lashings",
    "Orc-tusk lip piercings",
    "Umbrellas made from dead darkmantles",
    "Cloaks made of scraps from an elven tapestry",
];

const GOTHIC_TRINKETS: &[&str] = &[
    "A picture you drew as a child of your imaginary friend",
    "A lock that opens when blood is dripped in its keyhole",
    "Clothes stolen from a scarecrow",
    "A spinning top carved with four faces: happy, sad, wrathful, and dead",
    "The necklace of a sibling who died on the day you were born",
    "A wig from someone executed by beheading",
    "The unopened letter to you from your dying father",
    "A pocket watch that runs backward for an hour every midnight",
    "A winter coat stolen from a dying soldier",
    "A bottle of invisible ink that can only be read at sunset",
    "A wineskin that refills when interred with a dead person for a night",
    "A set of silverware used by a king for his last meal",
    "A spyglass that always shows the world suffering a terrible storm",
    "A cameo with the profile’s face scratched away",
    "A lantern with a black candle that never runs out and that burns with green flame",
    "A teacup from a child’s tea set, stained with blood",
    "A little black book that records your dreams, and yours alone, when you sleep",
    "A necklace formed of the interlinked holy symbols of a dozen deities",
    "A hangman’s noose that feels heavier than it should",
    "A birdcage into which small birds fly but once inside never eat or leave",
    "A lepidopterist’s box filled dead moths with skull-like patterns on their wings",
    "A jar of pickled ghouls’ tongues",
    "The wooden hand of a notorious pirate",
    "A urn with the ashes of a dead relative",
    "A hand mirror backed with a bronze depiction of a medusa",
    "Pallid leather gloves crafted with ivory fingernails",
    "Dice made from the knuckles of a notorious charlatan",
    "A ring of keys for forgotten locks",
    "Nails from the coffin of a murderer",
    "A key to the family crypt",
    "An bouquet of funerary flowers that always looks and smells fresh",
    "A switch used to discipline you as a child",
    "A music box that plays by itself whenever someone holding it dances",
    "A walking cane with an iron ferule that strikes sparks on stone",
    "A flag from a ship lost at sea",
    "Porcelain doll’s head that always seems to be looking at you",
    "A wolf’s head wrought in silver that is also a whistle",
    "A small mirror that shows a much older version of the viewer",
    "Small, worn book of children’s nursery rhymes",
    "A mummified raven claw",
    "A broken pendent of a silver dragon that’s always cold to the touch",
    "A small locked box that quietly hums a lovely melody at night but you always forget it in the morning",
    "An inkwell that makes one a little nauseous when staring at it",
    "An old little doll made from a dark, dense wood and missing a hand and a foot",
    "A black executioner’s hood",
    "A pouch made of flesh, with a sinew drawstring",
    "A tiny spool of black thread that never runs out",
    "A tiny clockwork figurine of a dancer that’s missing a gear and doesn’t work",
    "A black wooden pipe that creates puffs of smoke that look like skulls",
    "A vial of perfume, the scent of which only certain creatures can detect",
];
