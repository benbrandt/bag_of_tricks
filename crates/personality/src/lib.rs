use std::fmt;

use alignment::{AlignmentInfluences, Attitude, Morality};
use serde::{Deserialize, Serialize};
use strum::Display;

/// Types of alignment influence from personality traits
#[derive(Clone, Copy, Deserialize, Display, Serialize)]
pub enum Influence {
    Any,
    Chaotic,
    Evil,
    Good,
    Lawful,
    Neutral,
}

impl AlignmentInfluences for Influence {
    /// Return attitude influence from personality.
    /// Doubled because personality should be a major indicator of alignment.
    fn attitude(&self) -> Vec<Attitude> {
        match self {
            Self::Chaotic => vec![Attitude::Chaotic; 2],
            Self::Lawful => vec![Attitude::Lawful; 2],
            Self::Neutral => vec![Attitude::Neutral; 2],
            _ => vec![],
        }
    }

    /// Return morality influence from personality.
    /// Doubled because personality should be a major indicator of alignment.
    fn morality(&self) -> Vec<Morality> {
        match self {
            Self::Good => vec![Morality::Good; 2],
            Self::Evil => vec![Morality::Evil; 2],
            Self::Neutral => vec![Morality::Neutral; 2],
            _ => vec![],
        }
    }
}

/// Description of a character's personality
#[derive(Deserialize, Serialize)]
pub struct Personality {
    pub bond: String,
    pub flaw: String,
    pub ideal: (String, Influence),
    pub traits: Vec<String>,
}

/// Return attitude/morality influences from character's ideal
impl AlignmentInfluences for Personality {
    fn attitude(&self) -> Vec<Attitude> {
        self.ideal.1.attitude()
    }

    fn morality(&self) -> Vec<Morality> {
        self.ideal.1.morality()
    }
}

impl fmt::Display for Personality {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "PERSONALITY TRAITS:")?;
        for t in &self.traits {
            writeln!(f, "{}", t)?;
        }
        writeln!(f, "IDEAL: {} ({})", self.ideal.0, self.ideal.1)?;
        writeln!(f, "BOND: {}", self.bond)?;
        writeln!(f, "FLAW: {}", self.flaw)
    }
}

/// Trait to store associated constants for each background's personality tables.
pub trait PersonalityOptions {
    /// List of bonds to choose from
    fn bonds(&self) -> Vec<String> {
        vec![]
    }

    /// List of flaws to choose from
    fn flaws(&self) -> Vec<String> {
        vec![]
    }

    /// List of ideals to choose from, and their corresponding alignment influence
    fn ideals(&self) -> Vec<(String, Influence)> {
        vec![]
    }

    /// List of traits to choose from
    fn traits(&self) -> Vec<String> {
        vec![]
    }
}
