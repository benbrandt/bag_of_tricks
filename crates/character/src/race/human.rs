use std::fmt;

use backstory::Backstory;
use alignment::AlignmentInfluences;
use attack::Resistances;
use characteristics::{
    in_inches, names::Name, AgeRange, Appearance, CharacteristicDetails, Characteristics, Gender,
    HeightAndWeightTable, Size, Speed, WeightMod,
};
use citation::{Book, Citation, CitationList, Citations};
use dice_roller::{Die, RollCmd};
use languages::{Language, LanguageType, Languages};
use personality::PersonalityOptions;
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use trinkets::Trinkets;

use crate::{
    ability::{AbilityScore, AbilityScoreType},
    features::Features,
    proficiencies::Proficiencies,
};

use super::Race;

const HEIGHT_AND_WEIGHT: HeightAndWeightTable = HeightAndWeightTable {
    base_height: in_inches(4, 8),
    base_weight: 110,
    height_mod: RollCmd(2, Die::D10),
    weight_mod: WeightMod::Roll(RollCmd(2, Die::D4)),
};

#[derive(Deserialize, Serialize)]
pub(crate) struct Human;

impl Human {
    /// Separate function to make it easier to share with other races
    pub(crate) fn gen_first_name(
        rng: &mut impl Rng,
        CharacteristicDetails {
            ethnicity, gender, ..
        }: &CharacteristicDetails,
    ) -> &'static str {
        // Should be set if using this
        let names = ethnicity.unwrap().names();
        let first_names = match gender {
            Gender::Female => names.female,
            Gender::Male => names.male,
        };
        first_names.choose(rng).unwrap()
    }

    /// Separate function to make it easier to share with other races
    pub(crate) fn gen_surname(
        rng: &mut impl Rng,
        CharacteristicDetails { ethnicity, .. }: &CharacteristicDetails,
    ) -> &'static str {
        // Should be set if using this
        let names = ethnicity.unwrap().names();
        names.surname.choose(rng).unwrap_or(&"")
    }
}

impl AlignmentInfluences for Human {}

impl Appearance for Human {}

impl Backstory for Human {}

impl Characteristics for Human {
    const HUMAN_ANCESTRY: bool = true;
    const SIZE: Size = Size::Medium;

    fn get_age_range(&self) -> AgeRange {
        AgeRange(10..=100)
    }

    fn get_base_speeds(&self) -> Vec<Speed> {
        vec![Speed::Walking(30)]
    }

    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable {
        &HEIGHT_AND_WEIGHT
    }
}

impl Citations for Human {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Phb, 29)])
    }
}

impl Features for Human {}

impl Languages for Human {
    fn languages(&self) -> Vec<Language> {
        vec![Language::Common]
    }

    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (1, None)
    }
}

impl Name for Human {
    fn gen_name(rng: &mut impl Rng, characteristics: &CharacteristicDetails) -> String {
        format!(
            "{} {}",
            Self::gen_first_name(rng, characteristics),
            Self::gen_surname(rng, characteristics),
        )
    }
}

impl PersonalityOptions for Human {}

impl Proficiencies for Human {}

#[typetag::serde]
impl Race for Human {
    fn gen(rng: &mut impl Rng) -> (Box<dyn Race>, String, CharacteristicDetails) {
        let race = Box::new(Self);
        let characteristics = race.gen_characteristics(rng);
        let name = Self::gen_name(rng, &characteristics);
        (race, name, characteristics)
    }

    fn abilities(&self) -> Vec<AbilityScore> {
        AbilityScoreType::iter()
            .map(|t| AbilityScore(t, 1))
            .collect()
    }
}

impl Resistances for Human {}

impl Trinkets for Human {}

impl fmt::Display for Human {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Human")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let human = Human::gen(&mut rng);
        insta::assert_yaml_snapshot!(human);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let (human, _name, _characteristics) = Human::gen(&mut rng);
        insta::assert_snapshot!(format!("{}", human));
    }

    #[test]
    fn test_snapshot_abilities() {
        let human = Human;
        insta::assert_yaml_snapshot!(human.abilities());
    }

    #[test]
    fn test_snapshot_citations() {
        let human = Human;
        insta::assert_yaml_snapshot!(human.citations());
    }

    #[test]
    fn test_snapshot_features() {
        let human = Human;
        insta::assert_yaml_snapshot!(human.features());
    }
}
