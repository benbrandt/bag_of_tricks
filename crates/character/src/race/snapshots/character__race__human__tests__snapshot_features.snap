---
source: crates/character/src/race/human.rs
expression: human.features()

---
[]

